pkgs/dart_mcp/lib/src/server/resources_support.dart

```diff
@@ -0,0 +1,168 @@
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) async {
+    _peer.registerMethod(
+      ListResourcesRequest.methodName,
+      convertParameters(_listResources),
+    );
+
+    _peer.registerMethod(
+      ReadResourceRequest.methodName,
+      convertParameters(_readResource),
+    );
+
+    _peer.registerMethod(
+      SubscribeRequest.methodName,
+      convertParameters(_subscribeResource),
+    );
+    _peer.registerMethod(
+      UnsubscribeRequest.methodName,
+      convertParameters(_unsubscribeResource),
+    );
+
+    var result = await super.initialize(request);
+    (result.capabilities.resources ??= Resources())
+      ..listChanged = true
+      ..subscribe = true;
+    return result;
+  }
+
+  /// Register [resource] to call [impl] when invoked.
+  ///
+  /// If this server is already initialized and still connected to a client,
+  /// then the client will be notified that the resources list has changed.
+  ///
+  /// Throws a [StateError] if there is already a [Tool] registered with the
+  /// same name.
+  void addResource(
+    Resource resource,
+    FutureOr<ReadResourceResult> Function(ReadResourceRequest) impl,
+  ) {
+    if (_resources.containsKey(resource.uri)) {
+      throw StateError(
+        'Failed to add resource ${resource.name}, there is already a '
+        'resource that exists at the URI ${resource.uri}.',
+      );
+    }
+    _resources[resource.uri] = resource;
+    _resourceImpls[resource.uri] = impl;
+
+    if (ready) {
+      _notifyResourceListChanged();
+    }
+  }
+
+  /// Notifies the client that [resource] has been updated.
+  ///
+  /// The implementation of that resource can optionally be updated, otherwise
+  /// the previous implementation will be used.
+  ///
+  /// Throws a [StateError] if [resource] does not exist.
+  void updateResource(
+    Resource resource, {
+    FutureOr<ReadResourceResult> Function(ReadResourceRequest)? impl,
+  }) {
+    if (!_resources.containsKey(resource.uri)) {
+      throw StateError(
+        'Failed to update resource ${resource.name}, there is no resource '
+        'at the URI ${resource.uri}, you must add it first using addResource.',
+      );
+    }
+    if (impl != null) _resourceImpls[resource.uri] = impl;
+
+    if (_subscribedResources.contains(resource.uri)) {
+      _peer.sendNotification(
+        ResourceUpdatedNotification.methodName,
+        ResourceUpdatedNotification(uri: resource.uri),
+      );
+    }
+  }
+
+  /// Removes a [Resource] by [uri].
+  ///
+  /// Does not error if the resource hasn't been added yet.
+  void removeResource(String uri) {
+    _resources.remove(uri);
+    _resourceImpls.remove(uri);
+    if (ready) _notifyResourceListChanged();
+  }
+
+  /// Lists all the resources currently available.
+  ListResourcesResult _listResources(ListResourcesRequest request) {
+    return ListResourcesResult(
+      resources: [for (var resource in _resources.values) resource],
```

Comment:
can this just be `resources: _resources.values`?

Reply:
We need a List, but I did change it to just do `_resources.values.toList()`.